girder-dandi-archive/girder_dandi_archive/rest.py

```diff
@@ -69,21 +73,84 @@ def create_dandiset(self, params):
     @describeRoute(
         Description("Get Dandiset").param("identifier", "Dandiset Identifier", paramType="path")
     )
+    @dandiset_identifier
     def get_dandiset(self, identifier, params):
-
-        if not identifier:
-            raise RestException("identifier must not be empty.")
-
-        if not validate_dandiset_identifier(identifier):
-            raise RestException("Invalid Dandiset Identifier")
-
         # Ensure we are only looking for drafts collection child folders.
         drafts = get_or_create_drafts_collection()
         doc = Folder().findOne({"parentId": drafts["_id"], "meta.dandiset.identifier": identifier})
         if not doc:
             raise RestException("No such dandiset found.")
         return doc
 
+    @access.public
+    @describeRoute(
+        Description("Get Dandiset Owners").param(
+            "identifier", "Dandiset Identifier", paramType="path"
+        )
+    )
+    @dandiset_identifier
+    def get_dandiset_owners(self, identifier, params):
+        return Folder().getFullAccessList(find_dandiset_by_identifier(identifier))["users"]
+
+    @access.user
+    @autoDescribeRoute(
+        Description("Add Dandiset Owners")
+        .param("identifier", "Dandiset Identifier", paramType="path")
+        .jsonParam(
+            "owners",
+            "A JSON list of girder users to add as owners.",
+            paramType="body",
+            requireArray=True,
+        )
+    )
+    @dandiset_identifier
+    def add_dandiset_owners(self, identifier, owners, params):
+        dandiset = find_dandiset_by_identifier(identifier)
+        Folder().requireAccess(dandiset, user=self.getCurrentUser(), level=AccessType.ADMIN)
+
+        # Make sure the list doesn't contain duplicates
+        current_owners = [
+            (str(user["id"]), user["level"])
+            for user in Folder().getFullAccessList(dandiset)["users"]
+        ]
+        new_owners = [(user["_id"], AccessType.ADMIN) for user in owners]
+        final_owners = [
+            {"id": user_id, "level": level}
+            for user_id, level in set([*current_owners, *new_owners])
```

Comment:
If a user already has non-admin permissions on the folder and is added again with this method, that user will appear twice in the owners list with two different permissions levels.

Why not just use `user["_id"]` alone for `current_owners` and `new_owners` and do `final_owners = [{"id": user_id, "level": AccessType.ADMIN}]? This will upgrade any existing users to ADMIN, but that's currently what we expect 

If you're trying to preserve user permission levels, I would build a dict `{user_id: access_level}` of the current owners, add the new owners, then map that back to the required form of `[{"id":..., "level":...}]`. I think you did something similar in `remove_dandiset_owners`.